File: libs/authz-cedar/src/audit_kafka.rs
//! Kafka-backed [`AuthzAuditSink`] (feature `kafka`).
//!
//! Implements the production wiring sketched in [`crate::audit`]: every
//! authorization decision is serialised as JSON and published to
//! `audit.authz.v1` via the data-plane bus. Two design points worth
//! pinning down:
//!
//! * **Fire-and-forget.** [`AuthzAuditSink::emit`] returns immediately:
//!   the actual `produce()` call runs on a detached `tokio::spawn` task
//!   so a slow broker can never stall the request hot path. Failures are
//!   swallowed at `tracing::warn` level — the engine has already made
//!   its decision and the request must proceed.
//! * **Partition-by-principal.** The Kafka record key is the principal
//!   EntityUid string. This guarantees that all decisions for a given
//!   user land on the same partition, which keeps downstream sinks
//!   (Iceberg, audit-sink) able to reconstruct a per-user timeline
//!   without a global sort.
//!
//! `KafkaAuthzAuditSink` is built around `Arc<dyn DataPublisher>` rather
//! than the concrete `KafkaPublisher` so the test-suite can swap in a
//! capturing mock without spinning up `librdkafka`. Production callers
//! pass `Arc::new(KafkaPublisher::from_env(...)?) as Arc<dyn _>`.

use std::sync::Arc;

use async_trait::async_trait;
use event_bus_data::{DataPublisher, OpenLineageHeaders};
use uuid::Uuid;

use crate::audit::{AuthzAuditEvent, AuthzAuditSink};

/// Canonical Kafka topic for authorization decisions.
///
/// Provisioned in
/// `infra/k8s/platform/manifests/strimzi/topics-domain-v1.yaml`
/// (12 partitions, RF=3, ISR=2).
pub const TOPIC: &str = "audit.authz.v1";

/// OpenLineage `namespace` attached to every emitted record.
const OL_NAMESPACE: &str = "of://authz";

/// OpenLineage `job.name` — every emission belongs to the logical
/// "decide" job of the embedded Cedar engine.
const OL_JOB_NAME: &str = "authz.decide";

/// OpenLineage `producer` URI.
const OL_PRODUCER: &str = "https://github.com/unnamedlab/OpenFoundry/libs/authz-cedar";

/// Fire-and-forget Kafka audit sink.
#[derive(Clone)]
pub struct KafkaAuthzAuditSink {
    publisher: Arc<dyn DataPublisher>,
    topic: String,
}

impl KafkaAuthzAuditSink {
    /// Build a sink that publishes to `topic` via `publisher`.
    pub fn new(publisher: Arc<dyn DataPublisher>, topic: String) -> Self {
        Self { publisher, topic }
    }

    /// Convenience: publish to the canonical [`TOPIC`].
    pub fn for_default_topic(publisher: Arc<dyn DataPublisher>) -> Self {
        Self::new(publisher, TOPIC.to_string())
    }

    /// Topic this sink writes to (exposed for tests / metrics).
    pub fn topic(&self) -> &str {
        &self.topic
    }
}

#[async_trait]
impl AuthzAuditSink for KafkaAuthzAuditSink {
    async fn emit(&self, event: AuthzAuditEvent) {
        let publisher = Arc::clone(&self.publisher);
        let topic = self.topic.clone();
        tokio::spawn(async move {
            let payload = match serde_json::to_vec(&event) {
                Ok(bytes) => bytes,
                Err(err) => {
                    tracing::warn!(
                        target: "authz.audit.kafka",
                        error = %err,
                        principal = %event.principal,
                        action = %event.action,
                        "failed to serialise AuthzAuditEvent — dropping"
                    );
                    return;
                }
            };

            let headers = OpenLineageHeaders::new(
                OL_NAMESPACE,
                OL_JOB_NAME,
                Uuid::new_v4().to_string(),
                OL_PRODUCER,
            )
            .with_event_time(event.timestamp);

            let key = event.principal.as_bytes();
            if let Err(err) = publisher
                .publish(&topic, Some(key), &payload, &headers)
                .await
            {
                tracing::warn!(
                    target: "authz.audit.kafka",
                    error = %err,
                    topic = %topic,
                    principal = %event.principal,
                    action = %event.action,
                    decision = %event.decision,
                    "kafka publish failed for authz audit event — dropping"
                );
            }
        });
    }
}

File: openfoundry-go/libs/authz-cedar-go/audit_kafka.go
package cedarauthz

import (
	"context"
	"encoding/json"
	"log/slog"

	"github.com/google/uuid"

	databus "github.com/openfoundry/openfoundry-go/libs/event-bus-data"
)

// KafkaAuditTopic is the canonical topic for authorization decisions.
//
// Provisioned in
// `infra/k8s/platform/manifests/strimzi/topics-domain-v1.yaml`
// (12 partitions, RF=3, ISR=2).
const KafkaAuditTopic = "audit.authz.v1"

// OpenLineage facets attached to every emitted record. Locked to match
// the Rust impl byte-for-byte so downstream consumers (audit-trail,
// audit-compliance, audit-sink) see the same job/namespace/producer
// regardless of which runtime emitted the event.
const (
	olAuditNamespace = "of://authz"
	olAuditJobName   = "authz.decide"
	olAuditProducer  = "https://github.com/unnamedlab/OpenFoundry/libs/authz-cedar"
)

// KafkaAuthzAuditSink is the production [AuthzAuditSink].
//
// Two design points worth pinning down:
//
//   - **Fire-and-forget.** Emit returns immediately; the actual Publish
//     call runs on a detached goroutine so a slow broker can never stall
//     the request hot path. Failures are swallowed at WARN level.
//   - **Partition-by-principal.** The Kafka record key is the principal
//     EntityUID string. All decisions for a given user therefore land
//     on the same partition, which lets downstream sinks reconstruct a
//     per-user timeline without a global sort.
type KafkaAuthzAuditSink struct {
	publisher databus.Publisher
	topic     string
}

// NewKafkaAuditSink builds a sink that publishes to `topic` via
// `publisher`. Pass [KafkaAuditTopic] for the canonical topic.
func NewKafkaAuditSink(publisher databus.Publisher, topic string) *KafkaAuthzAuditSink {
	return &KafkaAuthzAuditSink{publisher: publisher, topic: topic}
}

// NewKafkaAuditSinkDefault is a convenience constructor wired to the
// canonical [KafkaAuditTopic].
func NewKafkaAuditSinkDefault(publisher databus.Publisher) *KafkaAuthzAuditSink {
	return NewKafkaAuditSink(publisher, KafkaAuditTopic)
}

// Topic returns the topic this sink writes to (exposed for tests / metrics).
func (s *KafkaAuthzAuditSink) Topic() string { return s.topic }

// Emit serialises the event to JSON and publishes it asynchronously.
//
// Implements [AuthzAuditSink].
func (s *KafkaAuthzAuditSink) Emit(ctx context.Context, event AuthzAuditEvent) {
	publisher := s.publisher
	topic := s.topic
	go func() {
		payload, err := json.Marshal(event)
		if err != nil {
			slog.Warn("authz.audit.kafka: failed to serialise AuthzAuditEvent — dropping",
				slog.String("error", err.Error()),
				slog.String("principal", event.Principal),
				slog.String("action", event.Action),
			)
			return
		}

		headers := databus.NewOpenLineageHeaders(
			olAuditNamespace, olAuditJobName,
			uuid.NewString(), olAuditProducer,
		).WithEventTime(event.Timestamp)

		key := []byte(event.Principal)
		if err := publisher.Publish(detachContext(ctx), topic, key, payload, &headers); err != nil {
			slog.Warn("authz.audit.kafka: kafka publish failed for authz audit event — dropping",
				slog.String("error", err.Error()),
				slog.String("topic", topic),
				slog.String("principal", event.Principal),
				slog.String("action", event.Action),
				slog.String("decision", event.Decision),
			)
		}
	}()
}
